itests/kit/node_unmanaged.go

```diff
@@ -464,332 +697,427 @@ func (tm *TestUnmanagedMiner) SubmitProveCommit(
 		Deadline:  sp.Deadline,
 		PoStIndex: 0,
 	}, 1, builtin.MethodsMiner.DisputeWindowedPoSt)
 	return err
 }
 
-func (tm *TestUnmanagedMiner) submitWindowPost(ctx context.Context, sectorNumber abi.SectorNumber, sealedCid cid.Cid, sealedPath, cacheDir string) error {
-	tm.t.Logf("Miner(%s): WindowPoST(%d): Running WindowPoSt ...\n", tm.ActorAddr, sectorNumber)
+func (tm *TestUnmanagedMiner) submitWindowPost(sectorNumbers []abi.SectorNumber) error {
+	if len(sectorNumbers) == 0 {
+		return fmt.Errorf("no sectors to submit window post for")
+	}
+
+	// We are limited to PoSting PoStedPartitionsMax (3) partitions at a time, so we need to group the
+	// sectors by partition and submit the PoSts in batches of PoStedPartitionsMax.
 
-	head, err := tm.FullNode.ChainHead(ctx)
+	partitionMap := make(map[uint64][]sectorInfo)
+	challengeEpoch := abi.ChainEpoch(-1) // should be the same for all
+
+	head, err := tm.FullNode.ChainHead(tm.ctx)
 	if err != nil {
 		return fmt.Errorf("Miner(%s): failed to get chain head: %w", tm.ActorAddr, err)
 	}
 
-	sp, err := tm.FullNode.StateSectorPartition(ctx, tm.ActorAddr, sectorNumber, head.Key())
+	di, err := tm.FullNode.StateMinerProvingDeadline(tm.ctx, tm.ActorAddr, head.Key())
 	if err != nil {
-		return fmt.Errorf("Miner(%s): failed to get sector partition for sector %d: %w", tm.ActorAddr, sectorNumber, err)
+		return fmt.Errorf("Miner(%s): failed to get proving deadline: %w", tm.ActorAddr, err)
 	}
 
-	di, err := tm.FullNode.StateMinerProvingDeadline(ctx, tm.ActorAddr, head.Key())
-	if err != nil {
-		return fmt.Errorf("Miner(%s): failed to get proving deadline for sector %d: %w", tm.ActorAddr, sectorNumber, err)
-	}
-	tm.t.Logf("Miner(%s): WindowPoST(%d): SectorPartition: %+v, ProvingDeadline: %+v\n", tm.ActorAddr, sectorNumber, sp, di)
-	if di.Index != sp.Deadline {
-		return fmt.Errorf("Miner(%s): sector %d is not in the deadline %d, but %d", tm.ActorAddr, sectorNumber, sp.Deadline, di.Index)
-	}
+	for _, sectorNumber := range sectorNumbers {
+		sector, err := tm.getCommittedSector(sectorNumber)
+		if err != nil {
+			return fmt.Errorf("Miner(%s): failed to get committed sector %d: %w", tm.ActorAddr, sectorNumber, err)
+		}
 
-	var proofBytes []byte
-	if tm.mockProofs {
-		proofBytes = []byte{0xde, 0xad, 0xbe, 0xef}
-	} else {
-		proofBytes, err = tm.generateWindowPost(ctx, sectorNumber, sealedCid, sealedPath, cacheDir)
+		sp, err := tm.FullNode.StateSectorPartition(tm.ctx, tm.ActorAddr, sectorNumber, head.Key())
 		if err != nil {
-			return fmt.Errorf("Miner(%s): failed to generate window post for sector %d: %w", tm.ActorAddr, sectorNumber, err)
+			return fmt.Errorf("Miner(%s): failed to get sector partition for sector %d: %w", tm.ActorAddr, sectorNumber, err)
+		}
+
+		if di.Index != sp.Deadline {
+			return fmt.Errorf("Miner(%s): sector %d is not in the deadline %d, but %d", tm.ActorAddr, sectorNumber, sp.Deadline, di.Index)
+		}
+
+		if challengeEpoch == abi.ChainEpoch(-1) {
```

Comment:
This does not need to be in this loop and should be pulled outside. Also, this if else does not make sense. They are not mutually exclusive conditions, right ?